src/theory/sets/theory_sets_rewriter.cpp

```diff
@@ -488,12 +488,12 @@ RewriteResponse TheorySetsRewriter::preRewrite(TNode node) {
   }
   else if (k == kind::INSERT)
   {
-    Node insertedElements = nm->mkNode(kind::SINGLETON, node[0]);
+    Node insertedElements = nm->mkSingleton(node[0].getType(), node[0]);
```

Comment:
This is (barely) safe I think, although it is fairly sketchy, think `(insert 1.0 1.5)`.  I believe this code works technically due to the fact this is at preRewrite before 1.0 from the parser (which is `(/ 1.0 1.0)` or `(to_real 1.0)`) is rewritten to 1.

You should compute the type you are going to use upfront (node[0].getType()) and then use assertions for each `node[i]` below that their types are subtypes of node[0].getType().

Reply:
Done. Funny I wrote `n[0].getType()` inside the loop, but then I looked at the type rule for insert, which uses type equality instead of subtyping. So I thought it is safe to use `n[i].getType()` here. 
Should we enable subtyping for the insert operator?